[macOS] background tab suspension - 3.5% MBA8,2 membuster regression

salinas-miguel opened the pull request:
#### 0420e99f106002c9da9e4803154ae715289a3087
<pre>
[macOS] background tab suspension - 3.5% MBA8,2 membuster regression
<a href="https://bugs.webkit.org/show_bug.cgi?id=255226">https://bugs.webkit.org/show_bug.cgi?id=255226</a>
rdar://106969950

Reviewed by Chris Dumez.

Membuster sends a low memory signal to web content processes to measure
the memory impact of our low memory handler. When runningboard
throttling is enabled on MBA8,2 the web content process never gets
scheduled due to the machine only having 2 cores and the web content
process having a priority of darwin_bg. This fix holds the foreground
assertion for 8 minutes after the last foreground activity is released
to ensure we have enough time to handle the low memory signal in
membuster. I also tried having the web content process take an assertion
on itself for the duration of the low memory handler but that did not
fix the regression.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottlerTimedActivity::ProcessThrottlerTimedActivity):
(WebKit::ProcessThrottlerTimedActivity::operator=):
(WebKit::ProcessThrottlerTimedActivity::activityTimedOut):
(WebKit::ProcessThrottlerTimedActivity::updateTimer):
(WebKit::ProcessThrottler::TimedActivity::TimedActivity): Deleted.
(WebKit::ProcessThrottler::TimedActivity::operator=): Deleted.
(WebKit::ProcessThrottler::TimedActivity::activityTimedOut): Deleted.
(WebKit::ProcessThrottler::TimedActivity::updateTimer): Deleted.
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::ProcessActivityState::ProcessActivityState):
(WebKit::WebPageProxy::ProcessActivityState::takeVisibleActivity):
(WebKit::WebPageProxy::ProcessActivityState::takeAudibleActivity):
(WebKit::WebPageProxy::ProcessActivityState::takeCapturingActivity):
(WebKit::WebPageProxy::ProcessActivityState::reset):
(WebKit::WebPageProxy::ProcessActivityState::dropVisibleActivity):
(WebKit::WebPageProxy::ProcessActivityState::dropAudibleActivity):
(WebKit::WebPageProxy::ProcessActivityState::dropCapturingActivity):
(WebKit::WebPageProxy::ProcessActivityState::hasValidVisibleActivity const):
(WebKit::WebPageProxy::ProcessActivityState::hasValidAudibleActivity const):
(WebKit::WebPageProxy::ProcessActivityState::hasValidCapturingActivity const):
(WebKit::WebPageProxy::ProcessActivityState::takeOpeningAppLinkActivity):
(WebKit::WebPageProxy::ProcessActivityState::dropOpeningAppLinkActivity):
(WebKit::WebPageProxy::ProcessActivityState::hasValidOpeningAppLinkActivity const):
(WebKit::WebPageProxy::close):
(WebKit::WebPageProxy::updateThrottleState):
(WebKit::WebPageProxy::clearAudibleActivity):
(WebKit::WebPageProxy::waitForDidUpdateActivityState):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::willOpenAppLink):

Canonical link: <a href="https://commits.webkit.org/263264@main">https://commits.webkit.org/263264@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5068918be2b4445aff935261676f02ccef5a638f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4032 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4131 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4243 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5478 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4283 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4226 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4107 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4405 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4084 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4273 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3639 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5479 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1774 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3615 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5797 "") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3599 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3676 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/5185 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4086 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3287 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3600 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3615 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1003 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3653 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3878 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -101,8 +107,22 @@ void ProcessThrottler::invalidateAllActivities()
 
+void ProcessThrottler::holdTemporaryForegroundAssertion()
+{
+    m_shouldHoldTemporaryForegroundAssertion = true;
```
Why a new boolean? Why can't we just hold a ProcessActivity in a data member like we usually do?
Maybe we'd hold it from 8 minutes after no longer being visible?

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -83,8 +85,12 @@ bool ProcessThrottler::addActivity(Activity& activity)
+    if (activity.isForeground()) {
         m_foregroundActivities.remove(&activity);
+        // If this was the last foreground activity keep holding the foreground assertion for a while
+        if (!m_foregroundActivities.size() && m_shouldDelaySuspension)
```
`m_foregroundActivities.isEmpty()`

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -83,8 +85,12 @@ bool ProcessThrottler::addActivity(Activity& activity)
         m_foregroundActivities.remove(&activity);
+        // If this was the last foreground activity keep holding the foreground assertion for a while
+        if (!m_foregroundActivities.size() && m_shouldDelaySuspension)
+            holdTemporaryForegroundAssertion();
```
Why do we need to hold a *foreground* assertion for a background tab. This seems a bit confusing. Why can't we hold a Background assertion?

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -42,7 +42,8 @@
     
 static const Seconds processSuspensionTimeout { 20_s };
-static const Seconds removeAllAssertionsTimeout { 8_min };
+static const Seconds removeAllAssertionsTimeout { 1_min };
```
This is confusing. This says we remove all assertions after 1 minute but the variable below says we're holding a temporary foreground assertion for 8 minutes.

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -42,7 +42,8 @@
 static const Seconds processSuspensionTimeout { 20_s };
-static const Seconds removeAllAssertionsTimeout { 8_min };
+static const Seconds removeAllAssertionsTimeout { 1_min };
+static const Seconds temporaryForegroundAssertionTimerTimeout { 8_min };
```
Temporary is not a clear enough name. We want a name that explains what this is for.

cdumez commented on Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
```diff
@@ -204,12 +204,17 @@ bool defaultMediaSessionCoordinatorEnabled()
 
 bool defaultRunningBoardThrottlingEnabled()
```
This seems to indicate that RunningBoard throttling is only enabled on macOS. Surely we do RunningBoard throttling on iOS too, no?

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -83,8 +85,12 @@ bool ProcessThrottler::addActivity(Activity& activity)
-    if (activity.isForeground())
+    if (activity.isForeground()) {
         m_foregroundActivities.remove(&activity);
+        // If this was the last foreground activity keep holding the foreground assertion for a while
```
Comments need to end with a period in WebKit, per coding style. Also, I think it would read better with a comma after "If this was the last foreground activity".

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.h:
```diff
@@ -115,9 +115,9 @@ class ProcessThrottler : public CanMakeWeakPtr<ProcessThrottler> {
+    bool shouldBeRunnable() const { return m_shouldHoldTemporaryForegroundAssertion || m_foregroundActivities.size() || m_backgroundActivities.size(); }
     void setAllowsActivities(bool);
-    void setShouldDropSuspendedAssertionAfterDelay(bool shouldDropAfterDelay) { m_shouldDropSuspendedAssertionAfterDelay = shouldDropAfterDelay; }
+    void setShouldDelaySuspension(bool delay) { m_shouldDelaySuspension = delay; }
```
delay -> shouldDelaySuspension

salinas-miguel commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2917,10 +2918,35 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
-    std::unique_ptr<ProcessThrottlerActivity> m_isCapturingActivity;
-    std::unique_ptr<ProcessThrottlerActivity> m_openingAppLinkActivity;
+
+    struct ProcessActivityState {
```
Should this go in WebPageProxyInternals.h? I wasn't sure.

cdumez replied:
The data members used to be here so I think it is fine to have the new struct/class in this header. However, I wouldn't declare the struct/class in the middle of data members. Please move the declaration before all data members.

salinas-miguel commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -460,6 +460,38 @@ Ref<WebPageProxy> WebPageProxy::create(PageClient& pageClient, WebProcessProxy&
     return adoptRef(*new WebPageProxy(pageClient, process, WTFMove(configuration)));
 }
 
+WebPageProxy::ProcessActivityState::ProcessActivityState(WebPageProxy& page)
```
I wasn't sure where to put this either. Maybe WebPageProxyInternals.h too?

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -460,6 +460,42 @@ Ref<WebPageProxy> WebPageProxy::create(PageClient& pageClient, WebProcessProxy&
+    , m_wasRecentlyVisibleActivity(makeUniqueRef<ProcessThrottlerTimedActivity>(8_min))
+{
+}
+void WebPageProxy::ProcessActivityState::takeVisibleActivity() { m_isVisibleActivity = m_page.process().throttler().foregroundActivity("View is visible"_s).moveToUniquePtr(); *m_wasRecentlyVisibleActivity = nullptr; }
```
Shouldn't be on a single line in cpp files. Also, please add a blank line between functions.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -460,6 +460,42 @@ Ref<WebPageProxy> WebPageProxy::create(PageClient& pageClient, WebProcessProxy&
 
+WebPageProxy::ProcessActivityState::ProcessActivityState(WebPageProxy& page)
+    : m_page(page)
+    , m_wasRecentlyVisibleActivity(makeUniqueRef<ProcessThrottlerTimedActivity>(8_min))
```
`#if PLATFORM(MAC)`

I think here we should make it clear that this behavior is platform-specific.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -510,6 +546,7 @@ WebPageProxy::WebPageProxy(PageClient& pageClient, WebProcessProxy& process, Ref
 #if ENABLE(FULLSCREEN_API)
     , m_fullscreenClient(makeUnique<API::FullscreenClient>())
 #endif
+    , m_processActivityState(makeUniqueRef<ProcessActivityState>(*this))
```
Do we really need to use a UniqueRef here? Grouping data members into their own class seems OK but I wonder what is the reasoning for heap allocating.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2917,10 +2918,39 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+    struct ProcessActivityState {
+        WTF_MAKE_STRUCT_FAST_ALLOCATED;
+
+        WebPageProxy& m_page;
```
Those data members are never used outside right? I think it would be nice to switch to a class and have the data members private.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2917,10 +2918,39 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+        WebPageProxy& m_page;
+
+        std::unique_ptr<ProcessThrottlerActivity> m_isVisibleActivity;
+        UniqueRef<ProcessThrottlerTimedActivity> m_wasRecentlyVisibleActivity;
```
I think we should guard this behind `PLATFORM(MAC)` since this is so macOS-specific.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2917,10 +2918,39 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+        UniqueRef<ProcessThrottlerTimedActivity> m_wasRecentlyVisibleActivity;
+        std::unique_ptr<ProcessThrottlerActivity> m_isAudibleActivity;
+        std::unique_ptr<ProcessThrottlerActivity> m_isCapturingActivity;
+        std::unique_ptr<ProcessThrottlerActivity> m_openingAppLinkActivity;
```
Isn't this iOS-specific? If it is, I think we should guard with IOS_FAMILY.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2917,10 +2918,39 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+    };
+
+    UniqueRef<ProcessActivityState> m_processActivityState;
+    ProcessActivityState& processActivityState() { return m_processActivityState; }
```
Getters should be in the middle of data members. Also not sure these getters are terribly useful. We could access m_processActivityState directly.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -28,6 +28,7 @@
 #include "APIObject.h"
 #include "MessageReceiver.h"
 #include "MessageSender.h"
+#include "ProcessThrottler.h"
```
Well, that is kind of defeating the purpose of extracting TimedActivity out of ProcessThrottler just so that you can forward declare it below.

salinas-miguel replied:
Is there a way to declare a `TimedActivity` property without the full definition from the header? I was using a `UniqueRef` to avoid this but I agree that a heap allocation is not optimal. If this is the only way I'll remove the forward declarations revert extracting `TimedActivity`

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2781,6 +2783,41 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+        std::unique_ptr<ProcessThrottlerActivity> m_openingAppLinkActivity;
+#endif
+
+    public:
```
These should come first and data members last. Please follow the style of the WebKit code base.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2781,6 +2783,41 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+#endif
+
+    public:
+        ProcessActivityState(WebPageProxy&);
```
explicit.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2781,6 +2783,41 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+
+        std::unique_ptr<ProcessThrottlerActivity> m_isVisibleActivity;
+#if PLATFORM(MAC)
+        ProcessThrottlerTimedActivity m_wasRecentlyVisibleActivity;
```
Why aren't you using a std::unique_ptr here like the others? Would solve your problem.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2917,10 +2954,8 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
-    std::unique_ptr<ProcessThrottlerActivity> m_isCapturingActivity;
-    std::unique_ptr<ProcessThrottlerActivity> m_openingAppLinkActivity;
+
+    ProcessActivityState m_processActivityState;
```
My earlier comment about not needed the unique_ptr only applied to this data member.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -460,6 +460,93 @@ Ref<WebPageProxy> WebPageProxy::create(PageClient& pageClient, WebProcessProxy&
+{
+    m_isVisibleActivity = m_page.process().throttler().foregroundActivity("View is visible"_s).moveToUniquePtr();
+#if PLATFORM(MAC)
+    *m_wasRecentlyVisibleActivity = nullptr;
```
Why the star? Shouldn't be needed.

cdumez replied:
Nevermind, bad leftover comment.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -460,6 +460,93 @@ Ref<WebPageProxy> WebPageProxy::create(PageClient& pageClient, WebProcessProxy&
+WebPageProxy::ProcessActivityState::ProcessActivityState(WebPageProxy& page)
+    : m_page(page)
+#if PLATFORM(MAC)
+    , m_wasRecentlyVisibleActivity(WTF::make_unique<ProcessThrottlerTimedActivity>(8_min))
```
`WTF::make_unique<>` -> `makeUniqueRef<>`.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2781,6 +2782,43 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+
+        std::unique_ptr<ProcessThrottlerActivity> m_isVisibleActivity;
+#if PLATFORM(MAC)
+        std::unique_ptr<ProcessThrottlerTimedActivity> m_wasRecentlyVisibleActivity;
```
Seems we could use a UniqueRef for this one given that it is never null.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2781,6 +2782,43 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+#if PLATFORM(IOS_FAMILY)
+        std::unique_ptr<ProcessThrottlerActivity> m_openingAppLinkActivity;
+#endif
+
```
nit: we should drop this blank line.

cdumez reviewed (approved):
LGTM with minor changes.

salinas-miguel committed:
[macOS] background tab suspension - 3.5% MBA8,2 membuster regression
https://bugs.webkit.org/show_bug.cgi?id=255226
rdar://106969950

Reviewed by Chris Dumez.

Membuster sends a low memory signal to web content processes to measure
the memory impact of our low memory handler. When runningboard
throttling is enabled on MBA8,2 the web content process never gets
scheduled due to the machine only having 2 cores and the web content
process having a priority of darwin_bg. This fix holds the foreground
assertion for 8 minutes after the last foreground activity is released
to ensure we have enough time to handle the low memory signal in
membuster. I also tried having the web content process take an assertion
on itself for the duration of the low memory handler but that did not
fix the regression.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottlerTimedActivity::ProcessThrottlerTimedActivity):
(WebKit::ProcessThrottlerTimedActivity::operator=):
(WebKit::ProcessThrottlerTimedActivity::activityTimedOut):
(WebKit::ProcessThrottlerTimedActivity::updateTimer):
(WebKit::ProcessThrottler::TimedActivity::TimedActivity): Deleted.
(WebKit::ProcessThrottler::TimedActivity::operator=): Deleted.
(WebKit::ProcessThrottler::TimedActivity::activityTimedOut): Deleted.
(WebKit::ProcessThrottler::TimedActivity::updateTimer): Deleted.
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::ProcessActivityState::ProcessActivityState):
(WebKit::WebPageProxy::ProcessActivityState::takeVisibleActivity):
(WebKit::WebPageProxy::ProcessActivityState::takeAudibleActivity):
(WebKit::WebPageProxy::ProcessActivityState::takeCapturingActivity):
(WebKit::WebPageProxy::ProcessActivityState::reset):
(WebKit::WebPageProxy::ProcessActivityState::dropVisibleActivity):
(WebKit::WebPageProxy::ProcessActivityState::dropAudibleActivity):
(WebKit::WebPageProxy::ProcessActivityState::dropCapturingActivity):
(WebKit::WebPageProxy::ProcessActivityState::hasValidVisibleActivity const):
(WebKit::WebPageProxy::ProcessActivityState::hasValidAudibleActivity const):
(WebKit::WebPageProxy::ProcessActivityState::hasValidCapturingActivity const):
(WebKit::WebPageProxy::ProcessActivityState::takeOpeningAppLinkActivity):
(WebKit::WebPageProxy::ProcessActivityState::dropOpeningAppLinkActivity):
(WebKit::WebPageProxy::ProcessActivityState::hasValidOpeningAppLinkActivity const):
(WebKit::WebPageProxy::close):
(WebKit::WebPageProxy::updateThrottleState):
(WebKit::WebPageProxy::clearAudibleActivity):
(WebKit::WebPageProxy::waitForDidUpdateActivityState):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::willOpenAppLink):

Canonical link: https://commits.webkit.org/263264@main
